### PR TITLE
Improved errors using type reference targets

### LINKED ISSUE #24776: Use type references for better error messages when relating to a union type
When relating `A<Foo>` to `A<Bar> | B<Baz> | C<Kwah>`, if the check fails, it is highly likely that the user wants to see the elaboration from `A<Foo>` to `A<Bar>`, rather than for `A<Foo>` to `B<Baz>` or `A<Foo>` to `C<Kwah>`.

We should try to find candidates when source types are type references, and target types are unions.

As a real-world example, here's an error message where TypeScript had the chance to relate `ComponentClass<...>` to `ComponentClass<...> | StatelessComponent<...>`:

https://news.ycombinator.com/item?id=17244638

image

**This should also work for if two types have a well-known type alias.**

Vaguely related to #6541?

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #24776

This pull request should improve error messages when assigning type references to union types when a target type contains a type reference whose target type matches the source's.

As an example,

```ts
interface A<T> {
    aProp: T;
}

interface B<T> {
    bProp: T;
}

interface C<T> {
    cProp: T;
}

declare const a: A<Foo>;
declare let thing: A<Bar> | B<Baz> | C<Kwah>;

thing = a;
```

Previously we would report

```
Type 'A<Foo>' is not assignable to type 'A<Bar> | B<Baz>'.
  Type 'A<Foo>' is not assignable to type 'B<Baz>'.
    Property 'bProp' is missing in type 'A<Foo>'.
```

Notice that we try to use `B<Baz>` to explain the incompatibility. Intuitively, we'd assume that `A<Bar>` is probably what the user wanted to assign to.

With this change, we report a more relevant error:

```
Type 'A<Foo>' is not assignable to type 'A<Bar> | B<Baz>'.
  Type 'A<Foo>' is not assignable to type 'A<Bar>'.
    Property 'bar' is missing in type 'Foo'.
```

~~Currently I'm trying to get this to also work for anonymous types which have a well-associated alias type, however, I might need help there (@weswigham @sandersn).~~

As future work, we can

* use inheritance chains of each type's target, since there may be cases when users relate a `Derived` to a `Base<Foo>`
* make this work when the source is an intersection types as well
* make this work on other type operators

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -10851,12 +10851,32 @@ namespace ts {
...
             }
 
+            function findMatchingTypeReferenceOrTypeAliasReference(source: Type, unionTarget: UnionOrIntersectionType) {
+                if (source.flags & TypeFlags.Object && (source as ObjectType).objectFlags & (ObjectFlags.Reference | ObjectFlags.Anonymous) && unionTarget.flags & TypeFlags.Union) {
+                    return find(unionTarget.types, target => {
+                        if (target.flags & TypeFlags.Object) {
+                            if ((source as ObjectType).objectFlags & (target as ObjectType).objectFlags & ObjectFlags.Reference) {
+                                return (source as TypeReference).target === (target as TypeReference).target;
+                            }
+                            if ((source as ObjectType).objectFlags & (target as ObjectType).objectFlags & ObjectFlags.Anonymous) {
```
Comment (weswigham) [inline src/compiler/checker.ts]: Use `getObjectFlags` instead of `(source as ObjectType).objectFlags`

**Comment by weswigham** [comment]
> It's odd that only one baseline changed besides the new test, and the change is not an improvement (although not much of a regression since it was quite confusing already). 

It's just selecting the first type instead of the last, since both are references to the tuple type.

**Comment by sandersn** [comment]
Oops, what I *really* meant to say was that it's odd that only one baseline changed. And also the one change isn't an improvement, which isn't odd because it's hard to change things without breaking something. Together, the two things make me wonder whether any more changes exist so that we can know whether they are improvements or not.

**Comment by DanielRosenwasser** [comment]
I agree; I was pretty disappointed to see the lack of improvement in the existing test suite. Maybe we'll see improvements in the RWC errors though.

**Pull request merged**